File: cronbach/cronbach.py
import pandas as pd
import numpy as np
import logging
import scipy.stats as stats
from typing import Dict, Union, Optional, List, Tuple
from dataclasses import dataclass

# Set up logging with more detailed configuration
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'
)

@dataclass
class CronbachResults:
    """Class to store Cronbach's alpha analysis results."""
    alpha: float
    item_total_correlations: pd.Series
    alpha_if_deleted: pd.Series
    confidence_interval: Tuple[float, float]
    std_error: float
    item_statistics: pd.DataFrame
    inter_item_correlations: pd.DataFrame
    scale_statistics: Dict[str, float]

def validate_data(df: pd.DataFrame, min_rows: int = 2, min_cols: int = 2) -> None:
    """
    Validate input data meets requirements for Cronbach's alpha calculation.
    
    Args:
        df: Input DataFrame to validate
        min_rows: Minimum required number of rows (participants)
        min_cols: Minimum required number of columns (items)
        
    Raises:
        TypeError: If input is not a DataFrame or contains non-numeric data
        ValueError: If data doesn't meet size requirements or has other issues
    """
    if not isinstance(df, pd.DataFrame):
        raise TypeError("Input must be a pandas DataFrame")
    
    if df.shape[0] < min_rows:
        raise ValueError(f"Data must have at least {min_rows} rows (participants)")
        
    if df.shape[1] < min_cols:
        raise ValueError(f"Data must have at least {min_cols} columns (items)")
        
    if df.isna().any().any():
        logging.warning("Data contains missing values which may affect results")
        
    if not all(np.issubdtype(dtype, np.number) for dtype in df.dtypes):
        raise TypeError("All columns must contain numeric data")
        
    # Check for constant columns
    if (df.std() == 0).any():
        logging.warning("One or more items have zero variance")
        
    # Check for reasonable value ranges
    if (df < 0).any().any():
        logging.warning("Data contains negative values - verify this is intended")

def calculate_confidence_interval(
    alpha: float, 
    n_items: int, 
    n_subjects: int, 
    confidence: float = 0.95
) -> Tuple[float, float]:
    """Calculate confidence interval for Cronbach's alpha."""
    # Using Fisher's transformation
    z = np.arctanh(alpha)
    se = np.sqrt(2 * (1 - alpha**2) / ((n_items - 1) * (n_subjects - 2)))
    z_crit = stats.norm.ppf((1 + confidence) / 2)
    
    lower = np.tanh(z - z_crit * se)
    upper = np.tanh(z + z_crit * se)
    
    return (lower, upper)

def calculate_item_statistics(df: pd.DataFrame) -> pd.DataFrame:
    """Calculate comprehensive item-level statistics."""
    stats_df = pd.DataFrame({
        'Mean': df.mean(),
        'Std': df.std(),
        'Skewness': df.skew(),
        'Kurtosis': df.kurtosis(),
        'Min': df.min(),
        'Max': df.max()
    })
    return stats_df

def cronbach_alpha(
    df: pd.DataFrame, 
    confidence: float = 0.95,
    handle_missing: str = 'pairwise'
) -> CronbachResults:
    """
    Calculate Cronbach's alpha and comprehensive reliability statistics.
    
    Args:
        df: pandas DataFrame with items as columns and participants as rows
        confidence: Confidence level for interval estimation (default: 0.95)
        handle_missing: Method for handling missing values ('pairwise', 'listwise', or 'impute')
        
    Returns:
        CronbachResults object containing:
        - alpha: Cronbach's alpha coefficient
        - item_total_correlations: Correlation of each item with total score
        - alpha_if_deleted: Alpha coefficient if each item is deleted
        - confidence_interval: Confidence interval for alpha
        - std_error: Standard error of alpha
        - item_statistics: Detailed item-level statistics
        - inter_item_correlations: Correlation matrix between items
        - scale_statistics: Overall scale statistics
    """
    try:
        validate_data(df)
        logging.info("Data validation passed. Processing comprehensive analysis...")
        
        # Handle missing values
        if handle_missing == 'listwise':
            df = df.dropna()
        elif handle_missing == 'impute':
            df = df.fillna(df.mean())
            
        # Number of items and participants
        N = df.shape[1]
        n_subjects = df.shape[0]
        
        # Variance calculations
        item_variances = df.var(axis=0, ddof=1)
        total_scores = df.sum(axis=1)
        var_total = total_scores.var(ddof=1)
        
        # Calculate alpha
        alpha = (N / (N - 1)) * (1 - item_variances.sum() / var_total)
        
        # Calculate confidence interval and standard error
        ci = calculate_confidence_interval(alpha, N, n_subjects, confidence)
        std_error = (ci[1] - ci[0]) / (2 * stats.norm.ppf((1 + confidence) / 2))
        
        # Item-total correlations
        item_total_correlations = pd.Series(
            {col: df[col].corr(total_scores - df[col]) for col in df.columns},
            name="Item-Total Correlations"
        )
        
        # Alpha if item deleted
        alpha_if_deleted = pd.Series(
            {col: cronbach_alpha(df.drop(columns=col))['alpha'] for col in df.columns},
            name="Alpha if Item Deleted"
        )
        
        # Calculate item statistics
        item_stats = calculate_item_statistics(df)
        
        # Calculate inter-item correlations
        inter_item_corr = df.corr()
        
        # Calculate scale statistics
        scale_stats = {
            'mean': total_scores.mean(),
            'variance': var_total,
            'std_dev': np.sqrt(var_total),
            'n_items': N,
            'n_subjects': n_subjects
        }
        
        results = CronbachResults(
            alpha=alpha,
            item_total_correlations=item_total_correlations,
            alpha_if_deleted=alpha_if_deleted,
            confidence_interval=ci,
            std_error=std_error,
            item_statistics=item_stats,
            inter_item_correlations=inter_item_corr,
            scale_statistics=scale_stats
        )
        
        return results
        
    except Exception as e:
        logging.error(f"Error calculating Cronbach's alpha: {str(e)}")
        raise

def display_results(results: CronbachResults, detailed: bool = True) -> None:
    """Display formatted results of the Cronbach's alpha analysis."""
    print("\nCRONBACH'S ALPHA ANALYSIS")
    print("=" * 70)
    
    print(f"\nOverall Cronbach's Alpha: {results.alpha:.3f}")
    print(f"95% Confidence Interval: ({results.confidence_interval[0]:.3f}, {results.confidence_interval[1]:.3f})")
    print(f"Standard Error: {results.std_error:.3f}")
    
    print("\nScale Statistics:")
    print("-" * 70)
    for key, value in results.scale_statistics.items():
        print(f"{key.replace('_', ' ').title()}: {value:.3f}")
    
    print("\nItem Analysis:")
    print("-" * 70)
    analysis = pd.DataFrame({
        'Item-Total Correlation': results.item_total_correlations,
        'Alpha if Item Deleted': results.alpha_if_deleted
    })
    print(analysis.round(3))
    
    if detailed:
        print("\nDetailed Item Statistics:")
        print("-" * 70)
        print(results.item_statistics.round(3))
        
        print("\nInter-Item Correlations:")
        print("-" * 70)
        print(results.inter_item_correlations.round(3))

# Example usage with error handling
if __name__ == "__main__":
    try:
        # Sample data
        data = {
            'Item1': [4, 3, 5, 2, 4, 5, 3, 2],
            'Item2': [3, 4, 2, 5, 3, 4, 4, 3],
            'Item3': [5, 4, 3, 4, 5, 3, 4, 4],
            'Item4': [2, 3, 4, 3, 4, 4, 3, 5]
        }
        
        df = pd.DataFrame(data)
        logging.info("Processing sample data...")
        
        results = cronbach_alpha(df)
        display_results(results)
        
    except Exception as e:
        logging.error(f"Program execution failed: {str(e)}")
        raise

File: cronbach/cronbach.m
function results = cronbach(data, varargin)
    % CRONBACH Compute comprehensive Cronbach's Alpha analysis
    %   results = CRONBACH(data) performs reliability analysis on input data
    %   results = CRONBACH(data, 'bootstrap', true) includes bootstrap CI
    %   results = CRONBACH(data, 'alpha', 0.05) sets confidence level (default 0.05)
    %
    % Input:
    %   data - numeric matrix where columns are items and rows are observations
    %   Name-Value pairs:
    %       'bootstrap' - logical, whether to compute bootstrap CI (default: false)
    %       'alpha' - numeric, significance level (default: 0.05)
    %       'nboot' - integer, number of bootstrap samples (default: 1000)
    %
    % Output:
    %   results - struct containing all reliability metrics and analysis
    
    % Parse inputs
    p = inputParser;
    addRequired(p, 'data', @isnumeric);
    addParameter(p, 'bootstrap', false, @islogical);
    addParameter(p, 'alpha', 0.05, @isnumeric);
    addParameter(p, 'nboot', 1000, @isnumeric);
    parse(p, data, varargin{:});
    
    % Validate data
    validateData(data);
    
    % Initialize results structure
    results = struct();
    
    % Basic sample characteristics
    results.n_observations = size(data, 1);
    results.n_items = size(data, 2);
    results.missing_values = sum(isnan(data(:)));
    
    % Compute Cronbach's Alpha
    [results.alpha_raw, results.alpha_std] = computeAlpha(data);
    
    % Item analysis
    results.item_statistics = computeItemStatistics(data);
    results.alpha_if_deleted = computeAlphaIfDeleted(data);
    results.inter_item_corr = corr(data, 'rows', 'pairwise');
    
    % Normality tests
    results.normality = checkNormality(data);
    
    % Bootstrap if requested
    if p.Results.bootstrap
        results.bootstrap = bootstrapAlpha(data, p.Results.nboot, p.Results.alpha);
    end
    
    % Display results if no output argument
    if nargout == 0
        displayResults(results);
    end
end

function validateData(data)
    % Input validation
    if size(data, 1) < 2
        error('Data must have at least 2 observations');
    end
    if size(data, 2) < 2
        error('Data must have at least 2 items');
    end
    if ~isnumeric(data)
        error('Data must be numeric');
    end
    if any(var(data) == 0)
        warning('Some items have zero variance');
    end
    if any(data(:) < 0, 'omitnan')
        warning('Data contains negative values - verify this is intended');
    end
end

function [alpha_raw, alpha_std] = computeAlpha(data)
    % Compute raw and standardized Cronbach's Alpha
    n_items = size(data, 2);
    
    % Raw alpha
    item_vars = var(data, 0, 'omitnan');
    total_var = var(sum(data, 2, 'omitnan'), 0);
    alpha_raw = (n_items/(n_items-1)) * (1 - sum(item_vars)/total_var);
    
    % Standardized alpha
    R = corr(data, 'rows', 'pairwise');
    mean_r = (sum(R(:)) - n_items) / (n_items^2 - n_items);
    alpha_std = (n_items * mean_r) / (1 + (n_items-1) * mean_r);
end

function stats = computeItemStatistics(data)
    % Compute detailed item statistics
    stats = struct();
    stats.means = mean(data, 'omitnan');
    stats.std = std(data, 0, 'omitnan');
    stats.min = min(data, [], 'omitnan');
    stats.max = max(data, [], 'omitnan');
    stats.skewness = skewness(data, 0);
    stats.kurtosis = kurtosis(data, 0) - 3; % Excess kurtosis
    stats.item_total_corr = computeItemTotalCorr(data);
end

function item_total_corr = computeItemTotalCorr(data)
    % Compute corrected item-total correlations
    n_items = size(data, 2);
    item_total_corr = zeros(1, n_items);
    
    for i = 1:n_items
        other_items = data(:, setdiff(1:n_items, i));
        total_score = sum(other_items, 2, 'omitnan');
        item_total_corr(i) = corr(data(:,i), total_score, 'rows', 'pairwise');
    end
end

function alpha_dropped = computeAlphaIfDeleted(data)
    % Compute alpha if item deleted
    n_items = size(data, 2);
    alpha_dropped = zeros(1, n_items);
    
    for i = 1:n_items
        reduced_data = data(:, setdiff(1:n_items, i));
        [alpha_dropped(i), ~] = computeAlpha(reduced_data);
    end
end

function normality = checkNormality(data)
    % Perform normality tests
    normality = struct();
    
    % Shapiro-Wilk test for each item
    n_items = size(data, 2);
    normality.shapiro_p = zeros(1, n_items);
    
    for i = 1:n_items
        [~, p] = swtest(data(:,i)); % Requires Statistics Toolbox
        normality.shapiro_p(i) = p;
    end
    
    % Basic multivariate normality indicators
    normality.mardia_skewness = maridia_skewness(data);
    normality.mardia_kurtosis = maridia_kurtosis(data);
end

function boot_results = bootstrapAlpha(data, n_boot, alpha)
    % Bootstrap confidence intervals for Cronbach's Alpha
    n_obs = size(data, 1);
    boot_alphas = zeros(n_boot, 1);
    
    for i = 1:n_boot
        % Sample with replacement
        boot_idx = randi(n_obs, n_obs, 1);
        boot_data = data(boot_idx, :);
        
        % Compute alpha for bootstrap sample
        [boot_alphas(i), ~] = computeAlpha(boot_data);
    end
    
    % Compute confidence intervals
    ci = [(alpha/2) (1-alpha/2)];
    boot_results.ci = quantile(boot_alphas, ci);
    boot_results.mean = mean(boot_alphas);
    boot_results.std = std(boot_alphas);
end

function displayResults(results)
    % Display comprehensive analysis results
    fprintf('\n=================================================\n');
    fprintf('COMPREHENSIVE RELIABILITY ANALYSIS RESULTS\n');
    fprintf('=================================================\n\n');
    
    % Main reliability metrics
    fprintf('RELIABILITY METRICS:\n');
    fprintf('--------------------------------------------------\n');
    fprintf('Cronbach''s Alpha (raw): %.3f\n', results.alpha_raw);
    fprintf('Cronbach''s Alpha (standardized): %.3f\n', results.alpha_std);
    
    if isfield(results, 'bootstrap')
        fprintf('Bootstrap 95%% CI: [%.3f, %.3f]\n', ...
            results.bootstrap.ci(1), results.bootstrap.ci(2));
    end
    
    % Sample characteristics
    fprintf('\nSAMPLE CHARACTERISTICS:\n');
    fprintf('--------------------------------------------------\n');
    fprintf('Sample size: %d observations\n', results.n_observations);
    fprintf('Number of items: %d\n', results.n_items);
    fprintf('Missing values: %d\n', results.missing_values);
    
    % Item statistics
    fprintf('\nITEM STATISTICS:\n');
    fprintf('--------------------------------------------------\n');
    fprintf('Item\tMean\tStd\tItem-Total Corr\tα if deleted\n');
    for i = 1:results.n_items
        fprintf('%d\t%.2f\t%.2f\t%.3f\t\t%.3f\n', ...
            i, ...
            results.item_statistics.means(i), ...
            results.item_statistics.std(i), ...
            results.item_statistics.item_total_corr(i), ...
            results.alpha_if_deleted(i));
    end
    
    % Interpretation guidelines
    fprintf('\nINTERPRETATION GUIDELINES:\n');
    fprintf('--------------------------------------------------\n');
    fprintf('Reliability Coefficient Interpretation:\n');
    fprintf('< 0.50: Unacceptable\n');
    fprintf('0.50 - 0.59: Poor\n');
    fprintf('0.60 - 0.69: Questionable\n');
    fprintf('0.70 - 0.79: Acceptable\n');
    fprintf('0.80 - 0.89: Good\n');
    fprintf('>= 0.90: Excellent\n');
end

function [H, pValue] = swtest(x)
    % Shapiro-Wilk test implementation
    % This is a simplified version - consider using Statistics Toolbox
    x = x(:);
    n = length(x);
    y = sort(x);
    
    % Compute W statistic
    m = norminv((1:n)' / (n+1));
    C = 1/sqrt(m'*m) * m;
    w = (C'*y)^2 / sum((y-mean(y)).^2);
    
    % Approximate p-value
    mu = -1.5861/sqrt(n);
    sigma = exp(0.8989 - 2.8605*n^(-1) + 2.4589*n^(-2));
    z = (log(1-w) - mu)/sigma;
    pValue = 1 - normcdf(z);
    H = (pValue < 0.05);
end

function b1p = maridia_skewness(X)
    % Mardia's multivariate skewness
    n = size(X,1);
    p = size(X,2);
    X = X - repmat(mean(X),n,1);
    S = cov(X);
    Sinv = inv(S);
    b1p = sum(sum((X*Sinv*X').^3))/n^2;
end

function b2p = maridia_kurtosis(X)
    % Mardia's multivariate kurtosis
    n = size(X,1);
    p = size(X,2);
    X = X - repmat(mean(X),n,1);
    S = cov(X);
    Sinv = inv(S);
    b2p = trace((X*Sinv*X').^2)/n;
end 